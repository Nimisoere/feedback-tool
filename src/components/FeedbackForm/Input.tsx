import React from 'react'
import { range } from '../../utils/common';
import { InputVariants, RenderInputProps } from './RenderInput';

const Input: React.FC<Omit<RenderInputProps, "label">> = React.forwardRef<HTMLInputElement & HTMLTextAreaElement, Omit<RenderInputProps, "label">>(
  function getInput({ id, error, inputType, variant, textareaProps, rangeInputProps, ...props }, ref) {
    const inputclassname = `${error ? "border-red-300 focus:outline-red-300" : "border-slate-300 focus:outline-cyan-400"} bg-slate-50 border-2 rounded-md py-2 px-4 `
    switch (variant) {
      case InputVariants.INPUT:
        return (
          <input id={id} ref={ref} type={inputType || "text"} className={inputclassname} {...props} />
        )
      case InputVariants.TEXT_AREA:
        return (
          <textarea id={id} ref={ref} className={inputclassname} {...textareaProps} {...props} />
        )
      case InputVariants.STAR_RATING: {
        const min = rangeInputProps?.min || 1;
        const max = rangeInputProps?.max || 5
        const step = rangeInputProps?.step || 1

        return (
          <div className='flex flex-wrap sm:flex-col'>
            {
              range(min, max, step)?.map((count) => (
                <div className='text-slate-400 w-full' key={`range-${count}-${id}`}>
                  <input ref={ref} id={`range-${count}-${id}`} className="hidden peer" type="radio" {...props} value={count} />
                  <label className='cursor-pointer text-sm peer peer-checked:text-yellow-400' htmlFor={`range-${count}-${id}`}>
                    <span className='text-2xl'>&#11089;</span> ({count} {count === 1 ? "star" : "stars"})
                  </label>
                </div>
              ))
            }
          </div>
        )
      }
      default:
        return (
          <input id={id} ref={ref} type={inputType || "text"} className={inputclassname} {...props} />
        )
    }
  })

export default Input;