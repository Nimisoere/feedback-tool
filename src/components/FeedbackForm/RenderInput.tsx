import React from 'react'
import { FieldError, Merge } from 'react-hook-form';
import Input from './Input';

export enum InputVariants {
  INPUT = "input",
  TEXT_AREA = "text-area",
  STAR_RATING = "star-rating"
}

type RangeInputProps = {
  min: number
  max: number
  step: number
}

export type RenderInputProps = {
  id: string;
  label: string | React.ReactNode;
  variant: InputVariants
  inputType?: React.HTMLInputTypeAttribute
  placeholder?: string;
  error?: Merge<FieldError, (FieldError | undefined)[]>
  textareaProps?: React.TextareaHTMLAttributes<HTMLTextAreaElement>
  rangeInputProps?: RangeInputProps
  ref?: React.ForwardedRef<HTMLInputElement & HTMLTextAreaElement>
}

const RenderInput: React.FC<RenderInputProps> = React.forwardRef(function InputRenderer({ label, error, ...props }, ref) {
  return (
    <div className='flex flex-col mb-8'>
      <label className='font-medium text-sm'>{label}</label>
      <Input error={error} ref={ref} {...props} />
      {error && <span className="text-red-900 text-sm font-medium">{error?.message}</span>}
    </div>
  )
})

export default RenderInput

