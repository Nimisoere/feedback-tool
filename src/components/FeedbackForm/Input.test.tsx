import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import Input from "./Input";
import { InputVariants } from "./RenderInput";

describe("Input control", () => {
  it("should match snapshot - Input", async () => {
    const { getByText, container } = render(
      <Input id="test" inputType={"email"} variant={InputVariants.INPUT} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should match snapshot - Rating", async () => {
    const { getByText, container } = render(
      <Input id="test"rangeInputProps={{ min: 1, max: 5, step: 1}} variant={InputVariants.STAR_RATING} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should match snapshot - Text Area", async () => {
    const { getByText, container } = render(
      <Input id="test" textareaProps={{ rows: 10 }} variant={InputVariants.TEXT_AREA} />
    );
    expect(container).toMatchSnapshot();
  });
});
