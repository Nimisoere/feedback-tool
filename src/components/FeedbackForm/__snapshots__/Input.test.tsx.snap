// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input control should match snapshot - Input 1`] = `
<div>
  <input
    class="border-slate-300 focus:outline-cyan-400 bg-slate-50 border-2 rounded-md py-2 px-4 "
    id="test"
    type="email"
  />
</div>
`;

exports[`Input control should match snapshot - Rating 1`] = `
<div>
  <div
    class="flex flex-wrap sm:flex-col"
  >
    <div
      class="text-slate-400 w-full"
    >
      <input
        class="hidden peer"
        id="range-1-test"
        type="radio"
        value="1"
      />
      <label
        class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
        for="range-1-test"
      >
        <span
          class="text-2xl"
        >
          ⭑
        </span>
         (
        1
         
        star
        )
      </label>
    </div>
    <div
      class="text-slate-400 w-full"
    >
      <input
        class="hidden peer"
        id="range-2-test"
        type="radio"
        value="2"
      />
      <label
        class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
        for="range-2-test"
      >
        <span
          class="text-2xl"
        >
          ⭑
        </span>
         (
        2
         
        stars
        )
      </label>
    </div>
    <div
      class="text-slate-400 w-full"
    >
      <input
        class="hidden peer"
        id="range-3-test"
        type="radio"
        value="3"
      />
      <label
        class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
        for="range-3-test"
      >
        <span
          class="text-2xl"
        >
          ⭑
        </span>
         (
        3
         
        stars
        )
      </label>
    </div>
    <div
      class="text-slate-400 w-full"
    >
      <input
        class="hidden peer"
        id="range-4-test"
        type="radio"
        value="4"
      />
      <label
        class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
        for="range-4-test"
      >
        <span
          class="text-2xl"
        >
          ⭑
        </span>
         (
        4
         
        stars
        )
      </label>
    </div>
    <div
      class="text-slate-400 w-full"
    >
      <input
        class="hidden peer"
        id="range-5-test"
        type="radio"
        value="5"
      />
      <label
        class="cursor-pointer text-sm peer peer-checked:text-yellow-400"
        for="range-5-test"
      >
        <span
          class="text-2xl"
        >
          ⭑
        </span>
         (
        5
         
        stars
        )
      </label>
    </div>
  </div>
</div>
`;

exports[`Input control should match snapshot - Text Area 1`] = `
<div>
  <textarea
    class="border-slate-300 focus:outline-cyan-400 bg-slate-50 border-2 rounded-md py-2 px-4 "
    id="test"
    rows="10"
  />
</div>
`;
